File: pkg/trajectory_ir/runtime/graft.py
"""Graft artifact references between trajectories (README §10 R07).

Transfers only artifact identity (content hash / ref metadata). Never copies
private THOUGHT nodes (or other private kinds) from source history.
"""

from __future__ import annotations

from typing import Any

from trajectory_ir.runtime.log import NodeLog
from trajectory_ir.runtime.nodes import Node

# Node kinds that must never cross a graft boundary.
PRIVATE_KINDS = frozenset({"THOUGHT"})

ARTIFACT_KINDS = frozenset({"ARTIFACT_PUT", "ARTIFACT_REF"})


class GraftError(Exception):
    """Raised when a graft request is invalid or would leak private data."""


def _payload_content_hash(payload: dict[str, Any]) -> str | None:
    for key in ("content_hash", "hash", "sha256"):
        v = payload.get(key)
        if isinstance(v, str) and v:
            return v
    return None


def find_artifact_ref(
    source_nodes: list[dict[str, Any]],
    content_hash: str,
) -> dict[str, Any]:
    """Locate ARTIFACT_PUT/REF metadata for content_hash; ignore private kinds."""
    if not content_hash or not isinstance(content_hash, str):
        raise GraftError("content_hash is required")
    # Reject accidental use of a thought id or non-hash token without search.
    matches: list[dict[str, Any]] = []
    for n in source_nodes:
        kind = n.get("kind")
        if kind in PRIVATE_KINDS:
            continue
        if kind not in ARTIFACT_KINDS:
            continue
        payload = n.get("payload") or {}
        if not isinstance(payload, dict):
            continue
        if _payload_content_hash(payload) == content_hash:
            matches.append(n)
    if not matches:
        raise GraftError(f"no ARTIFACT_PUT/ARTIFACT_REF with content_hash={content_hash!r}")
    # Prefer ARTIFACT_REF then PUT; stable by seq.
    matches.sort(key=lambda n: (0 if n.get("kind") == "ARTIFACT_REF" else 1, n.get("seq") or 0))
    return matches[0]


def graft_artifact_ref(
    target_log: NodeLog,
    *,
    content_hash: str,
    target_trajectory_id: str,
    target_tenant_id: str,
    seq: int,
    step_n: int | None = None,
    source_nodes: list[dict[str, Any]] | None = None,
    logical_path: str | None = None,
    uri: str | None = None,
    source_trajectory_id: str | None = None,
) -> Node:
    """Append an ARTIFACT_REF on the target log for ``content_hash``.

    If ``source_nodes`` is provided, metadata is taken from the first matching
    artifact node (never from THOUGHT). Private kinds are never copied as nodes.
    """
    if source_nodes is not None:
        # Fail closed if caller tries to pass a private-kind node as the only "match".
        for n in source_nodes:
            if n.get("kind") in PRIVATE_KINDS:
                ph = n.get("payload") if isinstance(n.get("payload"), dict) else {}
                if isinstance(ph, dict) and _payload_content_hash(ph) == content_hash:
                    raise GraftError("refusing to graft from a THOUGHT/private node")
        src = find_artifact_ref(source_nodes, content_hash)
        payload_src = src.get("payload") or {}
        if not isinstance(payload_src, dict):
            payload_src = {}
        logical_path = logical_path or payload_src.get("logical_path")  # type: ignore[assignment]
        uri = uri if uri is not None else payload_src.get("uri")  # type: ignore[assignment]
        if source_trajectory_id is None:
            source_trajectory_id = src.get("trajectory_id")  # type: ignore[assignment]

    payload: dict[str, Any] = {
        "content_hash": content_hash,
        "grafted": True,
    }
    if logical_path is not None:
        payload["logical_path"] = logical_path
    if uri is not None:
        payload["uri"] = uri
    if source_trajectory_id is not None:
        payload["source_trajectory_id"] = source_trajectory_id

    return target_log.append(
        "ARTIFACT_REF",
        step_n,
        payload,
        target_trajectory_id,
        target_tenant_id,
        seq,
    )

File: go/trajir/graft/graft.go
// Package graft transfers artifact refs between trajectories (R07).
// Never copies THOUGHT (or other private kinds) as nodes.
package graft

import (
	"fmt"
	"sort"

	nodelog "github.com/Coder-s-OG-s/Trajectory-IR/go/trajir/log"
	"github.com/Coder-s-OG-s/Trajectory-IR/go/trajir/nodes"
)

// PrivateKinds must never cross a graft boundary as full nodes.
var PrivateKinds = map[string]struct{}{
	"THOUGHT": {},
}

var artifactKinds = map[string]struct{}{
	"ARTIFACT_PUT": {},
	"ARTIFACT_REF": {},
}

// Error is a graft failure.
type Error struct {
	Msg string
}

func (e *Error) Error() string { return e.Msg }

func payloadContentHash(payload map[string]any) string {
	if payload == nil {
		return ""
	}
	for _, key := range []string{"content_hash", "hash", "sha256"} {
		if v, ok := payload[key].(string); ok && v != "" {
			return v
		}
	}
	return ""
}

// FindArtifactRef returns source ARTIFACT_* metadata for contentHash.
func FindArtifactRef(sourceNodes []map[string]any, contentHash string) (map[string]any, error) {
	if contentHash == "" {
		return nil, &Error{Msg: "content_hash is required"}
	}
	var matches []map[string]any
	for _, n := range sourceNodes {
		kind, _ := n["kind"].(string)
		if _, priv := PrivateKinds[kind]; priv {
			continue
		}
		if _, ok := artifactKinds[kind]; !ok {
			continue
		}
		payload, _ := n["payload"].(map[string]any)
		if payloadContentHash(payload) == contentHash {
			matches = append(matches, n)
		}
	}
	if len(matches) == 0 {
		return nil, &Error{Msg: fmt.Sprintf("no ARTIFACT_PUT/ARTIFACT_REF with content_hash=%q", contentHash)}
	}
	sort.SliceStable(matches, func(i, j int) bool {
		ki, _ := matches[i]["kind"].(string)
		kj, _ := matches[j]["kind"].(string)
		pi, pj := 1, 1
		if ki == "ARTIFACT_REF" {
			pi = 0
		}
		if kj == "ARTIFACT_REF" {
			pj = 0
		}
		if pi != pj {
			return pi < pj
		}
		si, sj := 0, 0
		if v, ok := matches[i]["seq"].(int); ok {
			si = v
		} else if v, ok := matches[i]["seq"].(float64); ok {
			si = int(v)
		}
		if v, ok := matches[j]["seq"].(int); ok {
			sj = v
		} else if v, ok := matches[j]["seq"].(float64); ok {
			sj = int(v)
		}
		return si < sj
	})
	return matches[0], nil
}

// GraftOptions configure GraftArtifactRef.
type GraftOptions struct {
	ContentHash         string
	TargetTrajectoryID  string
	TargetTenantID      string
	Seq                 int
	StepN               *int
	SourceNodes         []map[string]any
	LogicalPath         string
	URI                 string
	SourceTrajectoryID  string
	HasLogicalPath      bool
	HasURI              bool
	HasSourceTrajectory bool
}

// GraftArtifactRef appends ARTIFACT_REF on target. Never copies THOUGHT nodes.
func GraftArtifactRef(target *nodelog.NodeLog, opts GraftOptions) (*nodes.Node, error) {
	if target == nil {
		return nil, &Error{Msg: "nil NodeLog"}
	}
	logicalPath := opts.LogicalPath
	uri := opts.URI
	srcTraj := opts.SourceTrajectoryID
	if opts.SourceNodes != nil {
		for _, n := range opts.SourceNodes {
			kind, _ := n["kind"].(string)
			if _, priv := PrivateKinds[kind]; priv {
				payload, _ := n["payload"].(map[string]any)
				if payloadContentHash(payload) == opts.ContentHash {
					return nil, &Error{Msg: "refusing to graft from a THOUGHT/private node"}
				}
			}
		}
		src, err := FindArtifactRef(opts.SourceNodes, opts.ContentHash)
		if err != nil {
			return nil, err
		}
		payload, _ := src["payload"].(map[string]any)
		if !opts.HasLogicalPath {
			if v, ok := payload["logical_path"].(string); ok {
				logicalPath = v
			}
		}
		if !opts.HasURI {
			if v, ok := payload["uri"].(string); ok {
				uri = v
			}
		}
		if !opts.HasSourceTrajectory {
			if v, ok := src["trajectory_id"].(string); ok {
				srcTraj = v
			}
		}
	}
	payload := map[string]any{
		"content_hash": opts.ContentHash,
		"grafted":      true,
	}
	if logicalPath != "" || opts.HasLogicalPath {
		payload["logical_path"] = logicalPath
	}
	if uri != "" || opts.HasURI {
		payload["uri"] = uri
	}
	if srcTraj != "" || opts.HasSourceTrajectory {
		payload["source_trajectory_id"] = srcTraj
	}
	return target.Append(
		"ARTIFACT_REF",
		opts.StepN,
		payload,
		opts.TargetTrajectoryID,
		opts.TargetTenantID,
		opts.Seq,
	)
}
